File: web/backend/Python-services/src/models/UserAccount.py
"""
UserAccount model for FAMS
"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field
from ..constants import COLLECTIONS


class UserAccount(BaseModel):
    """UserAccount model - Base user representation in system"""
    userId: str = Field(..., description="Unique identifier for the user")
    email: str = Field(..., description="User's email address")
    password: str = Field(..., description="User's hashed password")
    role: str = Field(..., description="User role (admin, teacher, parent, student)")
    avatar: Optional[str] = Field(None, description="User's avatar URL")
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)
    isActive: bool = Field(default=True)
    collection: str = COLLECTIONS['USER_ACCOUNT'] 

File: web/backend/NodeJS-services/database/models/UserAccount.js
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');
const { COLLECTIONS, FIELDS } = require('../constants');

/**
 * UserAccount Schema
 * Represents users in the system with authentication capabilities
 */
const UserAccountSchema = new mongoose.Schema({
  userId: {
    type: String,
    required: true,
    unique: true
  },
  username: {
    type: String,
    required: true,
    unique: true
  },
  email: {
    type: String,
    required: true,
    unique: true
  },
  backup_email: {
    type: String,
    default: null
  },
  password: {
    type: String,
    required: true
  },
  role: {
    type: String,
    enum: ['admin', 'teacher', 'parent', 'student'],
    default: 'student'
  },
  avatar: {
    type: String,
    default: null
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true,
  versionKey: false,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

/**
 * Pre-save middleware to hash passwords
 * Only hashes the password if it has been modified
 */
UserAccountSchema.pre('save', async function(next) {
  if (!this.isModified('password')) {
    return next();
  }

  try {
    // Check if the password is already hashed
    if (this.password.startsWith('$2')) {
      return next();
    }

    const salt = await bcrypt.genSalt(12);
    this.password = await bcrypt.hash(this.password, salt);
    next();
  } catch (error) {
    console.error('Error hashing password:', error);
    next(error);
  }
});

/**
 * Method to match user entered password to hashed password in database
 */
UserAccountSchema.methods.matchPassword = async function(enteredPassword) {
  return await bcrypt.compare(enteredPassword, this.password);
};

module.exports = mongoose.model('UserAccount', UserAccountSchema, COLLECTIONS.USER_ACCOUNT); 